tools/engine_tool/lib/src/commands/run_command.dart

```diff
@@ -112,14 +113,22 @@ See `flutter run --help` for a listing
     return mode;
   }
 
+  Future<RunTarget?>? _runTarget;
+  Future<RunTarget?>? _getRunTarget() {
+    if (_runTarget != null) {
+      return _runTarget;
+    }
+    _runTarget = detectAndSelectRunTarget(environment, _getDeviceId());
+    return _runTarget;
+  }
```

Comment:
This is one of the use cases for `late final`, but I don't remember off the top of my head if you need to wrap it in a closure. It's either this:

```dart
  late final Future<RunTarget?> _runTarget = detectAndSelectRunTarget(environment, _getDeviceId());
```

Or this:
```dart
  late final Future<RunTarget?> _runTarget = (){
    return detectAndSelectRunTarget(environment, _getDeviceId());
  }();
```

Reply:
If it's a single expression I believe the former should work.

Reply:
Done, much nicer now!